rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
 //
 // The Client.Ready channel is closed after the client has connected
 // and completed one successful heartbeat. The Closed channel is
-// closed if the client fails to connect or if the client's Close()
-// method is invoked.
-func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
+// closed if the client's Close() method is invoked.
+func NewClient(addr net.Addr, context *Context) (*Client, error) {
 	clientMu.Lock()
+	defer clientMu.Unlock()
+
+	network := addr.Network()
+	address := addr.String()
+
+	key := network + address
```

Comment:
Mainly a theoretical concern, but this needs a delimiter to be unambiguous since some networks are prefixes of each other ("tcp" vs "tcp4" and "tcp6").